In [1]:
import pandas as pd

In [35]:
orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")

In [3]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype 
---  ------                         --------------  ----- 
 0   order_id                       99441 non-null  object
 1   customer_id                    99441 non-null  object
 2   order_status                   99441 non-null  object
 3   order_purchase_timestamp       99441 non-null  object
 4   order_approved_at              99281 non-null  object
 5   order_delivered_carrier_date   97658 non-null  object
 6   order_delivered_customer_date  96476 non-null  object
 7   order_estimated_delivery_date  99441 non-null  object
dtypes: object(8)
memory usage: 6.1+ MB


In [4]:
orders.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [7]:
orders.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [10]:
orders.order_status.unique()

array(['delivered', 'invoiced', 'shipped', 'processing', 'unavailable',
       'canceled', 'created', 'approved'], dtype=object)

In [17]:
99441 - 96478

2963

In [15]:
orders[(orders.order_status=='delivered') & (orders.order_delivered_customer_date.isna()==True)]

order_id                       customer_id  \
3002   2d1e2d5bf4dc7227b3bfebb81328c15f  ec05a6d8558c6455f0cbbd8a420ad34f   
20618  f5dd62b788049ad9fc0526e3ad11a097  5e89028e024b381dc84a13a3570decb4   
43834  2ebdfc4f15f23b91474edf87475f108e  29f0540231702fda0cfdee0a310f11aa   
79263  e69f75a717d64fc5ecdfae42b2e8e086  cfda40ca8dd0a5d486a9635b611b398a   
82868  0d3268bad9b086af767785e3f0fc0133  4f1d63d35fb7c8999853b2699f5c7649   
92643  2d858f451373b04fb5c984a1cc2defaf  e08caf668d499a6d643dafd7c5cc498a   
97647  ab7c89dc1bf4a1ead9d6ec1ec8968a84  dd1b84a7286eb4524d52af4256c0ba24   
98038  20edc82cf5400ce95e1afacc25798b31  28c37425f1127d887d7337f284080a0f   

      order_status order_purchase_timestamp    order_approved_at  \
3002     delivered      2017-11-28 17:44:07  2017-11-28 17:56:40   
20618    delivered      2018-06-20 06:58:43  2018-06-20 07:19:05   
43834    delivered      2018-07-01 17:05:11  2018-07-01 17:15:12   
79263    delivered      2018-07-01 22:05:55  2018-07-01 22:15:14   
82868    delivered      2018-07-01 21:14:02  2018-07-01 21:29:54   
92643    delivered      2017-05-25 23:22:43  2017-05-25 23:30:16   
97647    delivered      2018-06-08 12:09:39  2018-06-08 12:36:39   
98038    delivered      2018-06-27 16:09:12  2018-06-27 16:29:30   

      order_delivered_carrier_date order_delivered_customer_date  \
3002           2017-11-30 18:12:23                           NaN   
20618          2018-06-25 08:05:00                           NaN   
43834          2018-07-03 13:57:00                           NaN   
79263          2018-07-03 13:57:00                           NaN   
82868          2018-07-03 09:28:00                           NaN   
92643                          NaN                           NaN   
97647          2018-06-12 14:10:00                           NaN   
98038          2018-07-03 19:26:00                           NaN   

      order_estimated_delivery_date  
3002            2017-12-18 00:00:00  
20618           2018-07-16 00:00:00  
43834           2018-07-30 00:00:00  
79263           2018-07-30 00:00:00  
82868           2018-07-24 00:00:00  
92643           2017-06-23 00:00:00  
97647           2018-06-26 00:00:00  
98038           2018-07-19 00:00:00

In [24]:
orders[['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_delivered_customer_date']]

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp order_delivered_customer_date  
0        delivered      2017-10-02 10:56:33           2017-10-10 21:25:13  
1        delivered      2018-07-24 20:41:37           2018-08-07 15:27:45  
2        delivered      2018-08-08 08:38:49           2018-08-17 18:06:29  
3        delivered      2017-11-18 19:28:06           2017-12-02 00:28:42  
4        delivered      2018-02-13 21:18:39           2018-02-16 18:17:02  
...            ...                      ...                           ...  
99436    delivered      2017-03-09 09:54:05           2017-03-17 15:08:01  
99437    delivered      2018-02-06 12:58:58           2018-02-28 17:37:56  
99438    delivered      2017-08-27 14:46:43           2017-09-21 11:24:17  
99439    delivered      2018-01-08 21:28:27           2018-01-25 23:32:54  
99440    delivered      2018-03-08 20:57:30           2018-03-16 13:08:30  

[99441 rows x 5 columns]

In [23]:
len(orders) - len(orders[(orders.order_status=='delivered')])

2963

In [26]:
orders[(orders.order_status!='delivered')].isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 146
order_delivered_carrier_date     1781
order_delivered_customer_date    2957
order_estimated_delivery_date       0
dtype: int64

In [36]:
orders = orders[['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_delivered_customer_date']]

In [37]:
orders.order_purchase_timestamp = pd.to_datetime(orders.order_purchase_timestamp)
orders.order_delivered_customer_date = pd.to_datetime(orders.order_delivered_customer_date)

In [38]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  object        
 1   customer_id                    99441 non-null  object        
 2   order_status                   99441 non-null  object        
 3   order_purchase_timestamp       99441 non-null  datetime64[ns]
 4   order_delivered_customer_date  96476 non-null  datetime64[ns]
dtypes: datetime64[ns](2), object(3)
memory usage: 3.8+ MB


In [42]:
order_items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")

In [43]:
order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [45]:
order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [ ]:
order_items.shipping_limit_date = pd.to_datetime(order_items.shipping_limit_date)

In [47]:
orders

order_id                       customer_id  \
0      e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1      53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2      47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3      949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4      ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   
...                                 ...                               ...   
99436  9c5dedf39a927c1b2549525ed64a053c  39bd1228ee8140590ac3aca26f2dfe00   
99437  63943bddc261676b46f01ca7ac2f7bd8  1fca14ff2861355f6e5f14306ff977a7   
99438  83c1379a015df1e13d02aae0204711ab  1aa71eb042121263aafbe80c1b562c9c   
99439  11c177c8e97725db2631073c19f07b62  b331b74b18dc79bcdf6532d51e1637c1   
99440  66dea50a8b16d9b4dee7af250b4be1a5  edb027a75a1449115f6b43211ae02a24   

      order_status order_purchase_timestamp order_delivered_customer_date  
0        delivered      2017-10-02 10:56:33           2017-10-10 21:25:13  
1        delivered      2018-07-24 20:41:37           2018-08-07 15:27:45  
2        delivered      2018-08-08 08:38:49           2018-08-17 18:06:29  
3        delivered      2017-11-18 19:28:06           2017-12-02 00:28:42  
4        delivered      2018-02-13 21:18:39           2018-02-16 18:17:02  
...            ...                      ...                           ...  
99436    delivered      2017-03-09 09:54:05           2017-03-17 15:08:01  
99437    delivered      2018-02-06 12:58:58           2018-02-28 17:37:56  
99438    delivered      2017-08-27 14:46:43           2017-09-21 11:24:17  
99439    delivered      2018-01-08 21:28:27           2018-01-25 23:32:54  
99440    delivered      2018-03-08 20:57:30           2018-03-16 13:08:30  

[99441 rows x 5 columns]

In [67]:
orders_with_items = pd.merge(orders, order_items, on='order_id', how='left')

In [68]:
orders_with_items.shape

(113425, 11)

In [69]:
orders.shape

(99441, 5)

In [70]:
order_items.groupby('order_id')['order_item_id'].count().reset_index().sort_values('order_item_id', ascending=False)

order_id  order_item_id
50137  8272b63d03f5f79c56e9e4120aec44ef             21
10459  1b15974a0141d54e36626dca3fdc731a             20
65715  ab14fdcfbe524636d65ee38360e22ce8             20
60941  9ef13efd6949e4573a18964dd1bbe7f5             15
25583  428a2f660dc84138d969ccd69a0ab6d5             15
...                                 ...            ...
34640  5a0911d70c1f85d3bed0df1bf693a6dd              1
34639  5a082b558a3798d3e36d93bfa8ca1eae              1
34637  5a07264682e0b8fbb3f166edbbffc6e8              1
34636  5a071192a28951b76774e5a760c8c9b7              1
98665  fffe41c64501cc87c801fd61db3f6244              1

[98666 rows x 2 columns]

In [71]:
orders_with_items.duplicated().any()

np.False_

In [72]:
orders_with_items[orders_with_items.duplicated()==True]

Empty DataFrame
Columns: [order_id, customer_id, order_status, order_purchase_timestamp, order_delivered_customer_date, order_item_id, product_id, seller_id, shipping_limit_date, price, freight_value]
Index: []

### Summary
- There are 2963 cases where the orders were not delivered. Those are the cases where delivered_carrier_date, delivered_customer_Date, approved_At and mostly null expect a few. Only in 8 cases delivered_customer_date is null whenever order got delivered.  
- Dropped columns order_approved_at, order_delivered_carrier_date, order_estimated_delivery_date from orders table
- merged orders and order_items on order_id - left join
- previosuly order_id table has 99441 rows, after merge the orders_with_items table has 113425 rows. I have noticed there are multiple order_item_ids associated with each order_id so that can be the reason for increase in number